### PR TITLE
Backoff and token usage

### PR DESCRIPTION
**Pull request by dkolas**

### CODE REVIEW

File: src/api.py [hunk 1]
```diff
@@ -22,18 +32,30 @@ class PromptGenerationPlugin(Tagger):
...
 
     class PromptGenerationPluginConfig(Config):
-        openai_api_key: str = Field("", description="An openAI API key to use. If left default, will use Steamship's API key.")
+        openai_api_key: str = Field("",
+                                    description="An openAI API key to use. If left default, will use Steamship's API key.")
         max_words: int = Field(description="The maximum number of words to generate per request")
-        model: Optional[str] = Field("text-davinci-003", description="The OpenAI model to use.  Can be a pre-existing fine-tuned model.")
-        temperature: Optional[float] = Field(0.4, description="Controls randomness. Lower values produce higher likelihood / more predictable results; higher values produce more variety. Values between 0-1.")
-        top_p: Optional[int] = Field(1, description="Controls the nucleus sampling, where the model considers the results of the tokens with top_p probability mass. Values between 0-1.")
+        model: Optional[str] = Field("text-davinci-003",
```
Comment (EniasCailliau) [inline src/api.py:37]: Do we want to change this to a smaller model?

File: src/api.py [hunk 2]
```diff
@@ -42,59 +64,199 @@ def config_cls(cls) -> Type[Config]:
...
         super().__init__(client, config, context)
         openai.api_key = self.config.openai_api_key
 
+    def completion_with_retry(self, **kwargs: Any) -> Any:
+        """Use tenacity to retry the completion call."""
+        min_seconds = 4
+        max_seconds = 10
+
+        # Wait 2^x * 1 second between each retry starting with
+        # 4 seconds, then up to 10 seconds, then 10 seconds afterwards
```
Comment (douglas-reid) [inline src/api.py]: aside: i know this is a separate lib, but do you know if there is any jitter in the loop (post 10s retry)?

File: src/api.py [hunk 3]
```diff
@@ -42,59 +64,199 @@ def config_cls(cls) -> Type[Config]:
...
+            config: Dict[str, Any] = None,
+            context: InvocationContext = None,
     ):
         super().__init__(client, config, context)
         openai.api_key = self.config.openai_api_key
 
+    def completion_with_retry(self, **kwargs: Any) -> Any:
+        """Use tenacity to retry the completion call."""
+        min_seconds = 4
+        max_seconds = 10
```
Comment (douglas-reid) [inline src/api.py]: is the implication here (with default params) that _at most_ we expect up to 1 min of retrying in the `run` ?

File: src/api.py [hunk 4]
```diff
@@ -22,18 +32,30 @@ class PromptGenerationPlugin(Tagger):
     """
 
     class PromptGenerationPluginConfig(Config):
-        openai_api_key: str = Field("", description="An openAI API key to use. If left default, will use Steamship's API key.")
+        openai_api_key: str = Field("",
+                                    description="An openAI API key to use. If left default, will use Steamship's API key.")
         max_words: int = Field(description="The maximum number of words to generate per request")
```
Comment (EniasCailliau) [inline src/api.py:36]: Do we want to limit this by default? e.g. only 400 tokens unless you increase it yourself?

**Review by EniasCailliau — APPROVED** [review]
2 comments on setting less compute-heavy defaults

**Review by douglas-reid — APPROVED** [review]
LGTM. let's get this up and running!

File: src/api.py (hunk 2 above)
Comment (dkolas) [reply src/api.py]: I didn't look, maybe @EniasCailliau did?

**Pull request merged**